src/apps/MASS3DPA.hpp

```diff
@@ -0,0 +1,254 @@
+            for (int dz = 0; dz < D1D; ++dz) { \
+              RAJA_UNROLL(MQ1) \
+              for (int qz = 0; qz < Q1D; qz++) { \
+                u[qz] += DQQ[dz][qy][qx] * Bsmem[qz][dz]; \
+              } \
+            } \
+            RAJA_UNROLL(MQ1) \
+            for (int qz = 0; qz < Q1D; qz++) { \
+              QQQ[qz][qy][qx] = u[qz] * D_(qx, qy, qz, e); \
+            }
+
+#define MASS3DPA_6 \
+  Btsmem[d][q] = Bt_(q, d);
+
+#define MASS3DPA_7 \
+  double u[Q1D]; \
+RAJA_UNROLL(MQ1) \
+for (int qz = 0; qz < Q1D; ++qz) { \
+  u[qz] = 0; \
+ } \
+RAJA_UNROLL(MQ1) \
+for (int qx = 0; qx < Q1D; ++qx) { \
+  RAJA_UNROLL(MQ1) \
+    for (int qz = 0; qz < Q1D; ++qz) { \
+      u[qz] += QQQ[qz][qy][qx] * Btsmem[dx][qx]; \
+    } \
+ } \
+RAJA_UNROLL(MQ1) \
+for (int qz = 0; qz < Q1D; ++qz) { \
+  QQD[qz][qy][dx] = u[qz]; \
+ }
+
+#define MASS3DPA_8 \
+            double u[Q1D]; \
+            RAJA_UNROLL(MQ1) \
+            for (int qz = 0; qz < Q1D; ++qz) { \
+              u[qz] = 0; \
+            } \
+            RAJA_UNROLL(MQ1) \
+            for (int qy = 0; qy < Q1D; ++qy) { \
+              RAJA_UNROLL(MQ1) \
+              for (int qz = 0; qz < Q1D; ++qz) { \
+                u[qz] += QQD[qz][qy][dx] * Btsmem[dy][qy]; \
+              } \
+            } \
+            RAJA_UNROLL(MQ1) \
+            for (int qz = 0; qz < Q1D; ++qz) { \
+              QDD[qz][dy][dx] = u[qz]; \
+            }
+
+#define MASS3DPA_9 \
+            double u[D1D]; \
+            RAJA_UNROLL(MD1) \
+            for (int dz = 0; dz < D1D; ++dz) { \
+              u[dz] = 0; \
+            } \
+            RAJA_UNROLL(MQ1) \
+            for (int qz = 0; qz < Q1D; ++qz) { \
+              RAJA_UNROLL(MD1) \
+              for (int dz = 0; dz < D1D; ++dz) { \
+                u[dz] += QDD[qz][dy][dx] * Btsmem[dz][qz]; \
+              } \
+            } \
+            RAJA_UNROLL(MD1) \
+            for (int dz = 0; dz < D1D; ++dz) { \
+              Y_(dx, dy, dz, e) += u[dz]; \
+            }
+
+
+#if defined(RAJA_ENABLE_CUDA)
+  using device_launch = RAJA::expt::cuda_launch_t<true>;
+  using gpu_block_x_policy = RAJA::cuda_block_x_direct;
+  using gpu_thread_x_policy = RAJA::cuda_thread_x_loop;
+  using gpu_thread_y_policy = RAJA::cuda_thread_y_loop;
+#endif
+
+#if defined(RAJA_ENABLE_HIP)
+  using device_launch = RAJA::expt::hip_launch_t<true>;
+  using gpu_block_x_policy = RAJA::hip_block_x_direct;
+  using gpu_thread_x_policy = RAJA::hip_thread_x_loop;
+  using gpu_thread_y_policy = RAJA::hip_thread_y_loop;
+#endif
+
+namespace rajaperf
+{
+class RunParams;
+
+namespace apps
+{
+
+class MASS3DPA : public KernelBase
+{
+public:
+
+  MASS3DPA(const RunParams& params);
+
+  ~MASS3DPA();
+
+  void setUp(VariantID vid);
+  void updateChecksum(VariantID vid);
+  void tearDown(VariantID vid);
+
+  void runSeqVariant(VariantID vid);
+  void runOpenMPVariant(VariantID vid);
+  void runCudaVariant(VariantID vid);
+  void runHipVariant(VariantID vid);
+
```

Comment:
Have you tried compiling your branch with OpenMP target enabled? I think the runOpenMPTargetVariant method must be defined (it can be empty) here for the code to compile since it is pure virtual in KernelBase class.

Reply:
I'll check once I get access to the a machine.

Reply:
Good point, I just pushed some fixes. It also seems it needs the -OpenMP.cpp file. I wasn't able to get the code to build without it.